pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,422 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identitybackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	k8sTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "crd-allocator")
+)
+
+func NewCRDBackend(c CRDBackendConfiguration) (allocator.Backend, error) {
+	return &crdBackend{CRDBackendConfiguration: c}, nil
+}
+
+type CRDBackendConfiguration struct {
+	NodeName string
+	Store    cache.Store
+	Client   clientset.Interface
+	KeyType  allocator.AllocatorKey
+}
+
+type crdBackend struct {
+	CRDBackendConfiguration
+}
+
+func (c *crdBackend) DeleteAllKeys() {
+}
+
+// stripK8sPrefix strips the 'k8s:' prefix in the labels generated by
+// AllocatorKey.GetAsMap (when the key is k8s labels). In the CRD identity case
+// we map the labels directly to the ciliumidentity CRD instance, and
+// kubernetes does not allow ':' in the name of the label.
+func stripK8sPrefix(old map[string]string) map[string]string {
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		k = strings.TrimPrefix(k, "k8s:")
+		fixup[k] = v
+	}
+	return fixup
+}
+
+// AllocateID will create an identity CRD, thus creating the identity for this
+// key-> ID mapping.
+// Note: This does not create a reference to this node to indicate that it is
+// using this identity. That must be done with AcquireReference.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) AllocateID(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) error {
+	identity := &v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   id.String(),
+			Labels: stripK8sPrefix(key.GetAsMap()),
+		},
+		Status: v2.IdentityStatus{
+			Nodes: map[string]metav1.Time{
+				c.NodeName: metav1.Now(),
+			},
+		},
+	}
+
+	_, err := c.Client.CiliumV2().CiliumIdentities().Create(identity)
+	return err
+}
+
+func (c *crdBackend) AllocateIDIfLocked(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
+	return c.AllocateID(ctx, id, key)
+}
+
+// JSONPatch structure based on the RFC 6902
+type JSONPatch struct {
```

Comment:
JSONPatch structure already exists, `pkg/k8s/json_patch.go`

Reply:
discussed offline. Sadly, this causes an import loop :/ We'll leave this in for now but it may be possible to solve this and a few other loops by changing `pkg/identity/cache`. I think it imports k8s/identitybackend only to initialise the identity backend.
```
package github.com/cilium/cilium/cilium
        imports github.com/cilium/cilium/cilium/cmd
        imports github.com/cilium/cilium/pkg/bpf
        imports github.com/cilium/cilium/pkg/k8s
        imports github.com/cilium/cilium/pkg/ipcache
        imports github.com/cilium/cilium/pkg/identity/cache
        imports github.com/cilium/cilium/pkg/k8s/identitybackend
        imports github.com/cilium/cilium/pkg/k8s
```